### PR TITLE
G3tsmurf Misc updates

### PR DESCRIPTION
**Pull request by kmharrington**
This PR fixes a laundry list of issues:  #170 , #160, #137 , #173, #138, #135

Nothing should be breaking existing code. 

highlights:
1. Adds tag tracking to observations, automatically applying tags set at the time of taking the data
2. Adds some new generators that makes it easier to search through the SMuRF folder for things
3. Adds an "official" database update script into pipelines so we can standardize how that's called
4. Many bug fixes.

### CODE REVIEW

File: sotodlib/io/load_smurf.py [hunk 1]
```diff
@@ -1005,10 +1022,189 @@ def update_observation_files(self, obs, session, max_early=5, max_wait=100, forc
...
+                    continue
+                for root, dirs, files in os.walk(os.path.join(path, adir), topdown=False):
+                    for name in files:
+                        try:
+                            try:
+                                ctime = int(name.split('_')[0])
+                            except ValueError:
+                                ctime = actime
+                            fname = '_'.join(name.split('_')[1:])
+                            yield (fname, stream_id, ctime, os.path.join(root, name))
```
Comment (jlashner) [inline sotodlib/io/load_smurf.py:1119]: This method of returning ctimes + filenames seems strange to me, and doesn't really work if the first component of the filename is not a ctime (i.e. you will still be throwing away the first word). Maybe this should return the ctime and also the full filename including the first word?

File: sotodlib/io/load_smurf.py [hunk 2]
```diff
@@ -1468,12 +1616,25 @@ def from_time(cls, time, archive, show_pb=False):
...
                 Frames.type_name == 'Observation',
                 Frames.time <= time
-            ).order_by(Frames.time.desc()).first()
+            ).order_by(Frames.time.desc())
 
-        status_frames = session.query(Frames).filter(
+        if stream_id is not None:
+            q = q.join(Files).filter(Files.stream_id==stream_id)
+        else:
+            logger.warning(f"No stream_id given for finding status, not checking for one")
```
Comment (jlashner) [inline sotodlib/io/load_smurf.py:1627]: You can use the `_stream_ids_in_range` function to check if there are multiple stream-id options at a given time. I think if no stream_id is passed and there are multiple options this should raise an error.

File: pipelines/update_g3tsmurf_database.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    parser.add_argument('db_path', help="Path to Database to update")
+    
+    parser.add_argument('--timestream-folder', help="Absolute path to folder with .g3 timestreams. Overrides data_prefix")
+    parser.add_argument('--smurf-folder', help="Absolute path to folder with pysmurf archived data. Overrides data_prefix")
+    
+    parser.add_argument('--detdb-filename', help="File for dumping the context detector database")
+    
+    parser.add_argument('--update-delay', help="Days to subtract from now to set as minimum ctime",
+                               default=2, type=float)
+    parser.add_argument('--from_scratch', action="store_true")
```
Comment (jlashner) [inline pipelines/update_g3tsmurf_database.py:24]: `--from-scratch`

**Review by jlashner — APPROVED** [review]
Looks good to me! Just a few small comments inline but other than that I think this looks great, and I definitely appreciate the standardized update script.

File: sotodlib/io/load_smurf.py (hunk 1 above)
Comment (kmharrington) [reply sotodlib/io/load_smurf.py:1119]: I think the big thing here is the fname without the ctime makes for fast/easier pattern matching scripts. Matches similarly with the searching through the folders. Also, all the standard files from smurf all come with the ctime in front. I'd expect that to remain a requirement.

**Pull request merged**